packages/loot-core/src/server/budget/categoryTemplate.ts

```diff
@@ -0,0 +1,544 @@
+          break;
+        }
+        case 'by': {
+          //TODO add the logic to run all of these at once or whatever is needed
+          const ret = this.runBy(t[i], first, remainder);
+          toBudget += ret.ret;
+          remainder = ret.remainder;
+          first = false;
+          break;
+        }
+        case 'schedule': {
+          //TODO add this......
+          //TODO remember to trim the schedule name
+          const budgeted = getSheetValue(
+            monthUtils.sheetForMonth(this.month),
+            `leftover-${this.categoryID}`,
+          );
+          const ret = await goalsSchedule(
+            scheduleFlag,
+            t,
+            this.month,
+            budgeted,
+            remainder,
+            this.fromLastMonth,
+            toBudget,
+            [],
+            this.categoryID,
+          );
+          toBudget = ret.to_budget;
+          remainder = ret.remainder;
+          scheduleFlag = ret.scheduleFlag;
+          break;
+        }
+        case 'average': {
+          toBudget += await this.runAverage(t[i]);
+          break;
+        }
+      }
+
+      // don't overbudget when using a priority
+      if (priority > 0 && toBudget > available) {
+        toBudget = available;
+      }
+      available = available - toBudget;
+
+      if (priority > 0 && available <= 0) {
+        break;
+      }
+    }
+
+    this.toBudgetAmount += toBudget;
+    return toBudget;
+  }
+
+  applyLimit(): number {
+    if (this.limitCheck === false) {
+      return 0;
+    }
+    if (this.limitHold && this.fromLastMonth >= this.limitAmount) {
+      const orig = this.toBudgetAmount;
+      this.toBudgetAmount = 0;
+      return orig;
+    }
+    if (this.toBudgetAmount + this.fromLastMonth > this.limitAmount) {
+      const orig = this.toBudgetAmount;
+      this.toBudgetAmount = this.limitAmount - this.fromLastMonth;
+      return orig - this.toBudgetAmount;
+    }
+    return 0;
+  }
+
+  // run all of the 'remainder' type templates
+  runRemainder(budgetAvail: number, perWeight: number) {
+    if (this.remainder.length === 0) return 0;
+    const toBudget = Math.round(this.remainderWeight * perWeight);
+    //check possible overbudget from rounding, 1cent leftover
+    if (toBudget > budgetAvail) {
+      this.toBudgetAmount += budgetAvail;
+    } else if (budgetAvail - toBudget === 1) {
+      this.toBudgetAmount += toBudget + 1;
+    } else {
+      this.toBudgetAmount += toBudget;
+    }
+    return toBudget;
+  }
+
+  getValues() {
+    this.runGoal();
+    return {
+      budgeted: this.toBudgetAmount,
+      goal: this.goalAmount,
+      longGoal: this.isLongGoal,
+    };
+  }
+
+  //-----------------------------------------------------------------------------
+  // Implimentation
+  readonly categoryID: string; //readonly so we can double check the category this is using
+  private month: string;
+  private templates = [];
+  private remainder = [];
+  private goals = [];
```

Comment:
As this can only legally be 0 or 1. Why no use:

```suggestion
  private goal?: GoalTemplate;
```